### PR TITLE
Adding Vitess to the marketplace

### PR DESCRIPTION
**Pull request by dkhenry**
## BACKGROUND
This change Adds the Vitess Operator and a simple example cluster to the marketplace. Vitess is a distributed database system that provides a mysql compatibility while giving the user the ability to scale out their datastore through sharding. It is a graduated CNCF project used in production by a number of companies. More information can be found at https://vitess.io
-----------------------------------------------------------------------
## Changes
* Add in the planetscale vitess-operator and an example deployment

-----------------------------------------------------------------------

## Checklist
- [x] review the [contributing doc](https://github.com/digitalocean/marketplace-kubernetes/blob/master/CONTRIBUTING.md) with steps for both adding or updating your application (if applicable)
------------------------------------------------------------------------

Reviewer: @marketplace-eng

### CODE REVIEW

**Comment by scottcrawford03** [comment]
👋 thanks for submitting. were you able to get this running on a DigitalOcean k8s cluster using deploy-local.sh? I have been getting timeout errors and failure to connect to topo server errors. It looks as though the etcd pods are stuck in pending.

**Comment by scottcrawford03** [comment]
I was able to get it to run through. the gotcha was it required 6 volume claims. In the Readme we need to make sure that we call out resource minimums for this otherwise it's going to seems like it's just failing like it was for me.

Also, i'm getting timeouts from this line here:
`kubectl wait --for=condition=Ready --timeout=1m vitesscluster/"${line}"`

after digging in some more it looks like there isn't a Condition section on the VitessCluster resource. Is there something else that we should be looking at to determine the readiness of a VitessCluster resource?

**Comment by dkhenry** [comment]
Updated for the comments you left, and I added a README.md This should now work with just 4 PVC's and even fewer resources

File: stacks/vitess/README.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
+# Getting Started with the Vitess Operator
+
+After you have deployed this stack you will have the Vitess operator and an
```
Comment (bnlandry) [inline stacks/vitess/README.md:3]: It looks like we want to have a section in this README explaining what the Vitess Operator does and its value to the user. Also, the template includes a boilerplate section on getting K8s running on DO. See https://github.com/digitalocean/marketplace-kubernetes/blob/master/docs/LISTING-README-TEMPLATE.md

File: stacks/vitess/README.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+example Database cluster. 
+
+## Base Requirements
+The example database requires four persistent volume claims ( 3 for the
+lockserver, and 1 for the database ). It also requires three nodes to function
+effectively. The operator will make its best effort to spread all components
+across nodes to ensure availability. 
+
+## Accessing the Cluster
+To log into this cluster you can use the following commands
```
Comment (bnlandry) [inline stacks/vitess/README.md:13]: Suggestion: "To log into the example database cluster, use the following commands:"

File: stacks/vitess/README.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+
+### Access the Database from outside the Kubernetes cluster
+You can access the database outside the kubernetes cluster by port-forwarding to
```
Comment (bnlandry) [inline stacks/vitess/README.md:17]: Please capitalize "Kubernetes"

File: stacks/vitess/README.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+the vtgate service. This service will load balance connection to all vtgates.
```
Comment (bnlandry) [inline stacks/vitess/README.md:18]: "connections"?

File: stacks/vitess/README.md [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
```
Comment (bnlandry) [inline stacks/vitess/README.md:19]: Try introducing these commands: "Use the following commands to port-forward to the vtgate service:"

File: stacks/vitess/README.md [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+``` sh
+VTGATE=$(kubectl get service -l "planetscale.com/component=vtgate,planetscale.com/cluster=example" -o custom-columns=Name:.metadata.name | tail -n -1)
+kubectl port-forward service/"${VTGATE}" 3306&
+```
+
+you will then be able to access your vitess cluster on the forwarded port with
```
Comment (bnlandry) [inline stacks/vitess/README.md:25]: Suggestion: "You can now access your Vitess cluster on the forwarded port with
the standard MySQL client."

File: stacks/vitess/README.md [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+the standard mysql client 
+
+``` sh
+mysql -h 127.0.0.1 -u admin -pscalewithvitess main 
+```
+
+### Access the Database from inside the Kubernetes cluster
```
Comment (bnlandry) [inline stacks/vitess/README.md:32]: For consistency, capitalize "Cluster"

File: stacks/vitess/README.md [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+Inside the kubernetes cluster the vtgate service will be accessable on the
```
Comment (bnlandry) [inline stacks/vitess/README.md:33]: Suggestions: "Inside the Kubernetes cluster, the vtgate service will be accessible on the
service DNS name. You can find all vtgate services using the following command:"

File: stacks/vitess/README.md [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+service DNS name. You can find all vtgate services by running 
+
+``` sh
+kubectl get service -l "planetscale.com/component=vtgate,planetscale.com/cluster=example"
+```
+
+Vtgate exposes a mysql compatible server on port 3306. It also has a gRPC server
```
Comment (bnlandry) [inline stacks/vitess/README.md:40]: For consistency: "The vtgate server..."

File: stacks/vitess/README.md [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+on port 15999
```
Comment (bnlandry) [inline stacks/vitess/README.md:41]: Punctuation: "... on port 15999."

File: stacks/vitess/README.md [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+
+
+# Customizing the Database
```
Comment (bnlandry) [inline stacks/vitess/README.md:44]: Suggestion: "...from this location:"

File: stacks/vitess/README.md [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+The default database deployed by this stack is an example database that is not
+suitable for production workloads. You can customize the example database by
+fetching the YAML manifest from here 
+
+``` sh
+https://raw.githubusercontent.com/digitalocean/marketplace-kubernetes/master/stacks/vitess/yaml/example-cluster.yaml
+```
+
+## Important Changes for Production workloads
```
Comment (bnlandry) [inline stacks/vitess/README.md:53]: For consistency: "...for Production Workloads"

File: stacks/vitess/README.md [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+General recommendations for production workloads can be found on the main Vitess
```
Comment (bnlandry) [inline stacks/vitess/README.md:54]: Suggestion: "The Vitess documentation contains [general recommendations for production workloads](https://vitess.io/docs/user-guides/production-planning/):

File: stacks/vitess/README.md [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+documentation site 
+
+https://vitess.io/docs/user-guides/production-planning/
+
+
+### Username and Password
+In the example manifest there is a kubernetes secret named
```
Comment (bnlandry) [inline stacks/vitess/README.md:61]: Suggestion: "The example manifest contains a Kubernetes secret named
`example-cluster-config`. That secret contains the username and password of the
default user. Change these values to a secure username and password.
See the Vitess documentation for [how to configure users and passwords](https://vitess.io/docs/user-guides/user-management/):"

File: stacks/vitess/README.md [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+`example-cluster-config`. That secret contains the username as password of the
+default user. This should be changed to a secure username and password.
+Reference the vitess documentation on how to configure users and passwords
+
+https://vitess.io/docs/user-guides/user-management/
+
+### Number of Gateways
+By default only one vtgate pod is created. To ensure reliability this should be
```
Comment (bnlandry) [inline stacks/vitess/README.md:69]: Suggestion: "By default, this example creates only one vtgate pod. To ensure reliability, increase the number of vtgate pods to two or more. This setting is found here:"

File: stacks/vitess/deploy-local.sh [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+kubectl get vitesskeyspace -o custom-columns=NAME:.metadata.name | tail -n +2 | while read -r line
+do
+  # Since a vitesskeyspaces doesn't support conditions we have to poll the master status
+  retry=0
+  while [ "${retry}" -lt 30 ] && [ "$(kubectl get vitesskeyspace/"${line}" -o=jsonpath='{.status.shards..hasMaster}' | xargs -n1 echo | sort -u )" != "True" ]; do
+    echo "Waiting for vitess keyspace ${line}"
+    retry=$(( retry+1 ))
+    sleep 10
+  done
+  echo "keyspce ${line} is ready."
```
Comment (scottcrawford03) [inline stacks/vitess/deploy-local.sh:36]: nit: typo on keyspace

File: stacks/vitess/deploy.sh [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+kubectl get vitesskeyspace -o custom-columns=NAME:.metadata.name | tail -n +2 | while read -r line
+do
+  # Since a vitesskeyspaces doesn't support conditions we have to poll the master status
+  retry=0
+  while [ "${retry}" -lt 30 ] && [ "$(kubectl get vitesskeyspace/"${line}" -o=jsonpath='{.status.shards..hasMaster}' | xargs -n1 echo | sort -u )" != "True" ]; do
+    echo "Waiting for vitess keyspace ${line}"
+    retry=$(( retry+1 ))
+    sleep 10
+  done
+  echo "keyspce ${line} is ready."
```
Comment (scottcrawford03) [inline stacks/vitess/deploy.sh:36]: nit: typo on keyspce

**Comment by scottcrawford03** [comment]
Looks like there's a lot of README updates requested. Everything else LGTM. I'll 🍏when the README is updated and I look over that once more 👍

**Comment by scottcrawford03** [comment]
LGTM.

@dkhenry can you fill out the google form here: https://docs.google.com/forms/d/1fzbOUYTMXbrn9cECEZA6ya0zITcdCWtpJWHfkxmsWAQ

this will give us all the details we need to create the listing on the marketplace. thanks!

**Pull request merged**